### PR TITLE
Improve IDL 1 and 2 interoperability

### PR DESCRIPTION
**Pull request by mtdowling**
This commit makes several significant updates to Smithy IDL 2 that
improve lossless interoperability with Smithy IDL 1:

* Default traits can now coexist with required. This indicates that
  a member should be serialized, but it is a protocol-specific
  decision if and how this is enforced. This was a pattern that occurred
  in Smithy 1.0 models when a member was required and targeted a shape
  with a zero value.
* Default traits can be added to root-level shapes. Any structure member
  that targets a shape marked with the default trait must repeat the
  default on the member. This removes the action at a distance problem
  observed in Smithy IDL 1 where a root level shape implicitly introduced
  a default zero value, and to know if that's the case for any member, you
  had to look through from the member to the target shape. This change
  allows us to know if a root level shape was boxed in IDL 1.0 too -- if
  it is a shape that had a zero value in IDL 1 and either doesn't have
  a default or the default is not the zero value, then it was boxed in v1.
* When loading v2 models, we will now patch synthetic box traits onto shapes
  that would have been considered boxed by existing smithy-model consumers.
  This improves further interop with tooling that has not yet adopted
  Smithy IDL 2 or that hasn't yet migrated to use the NullableIndex
  abstraction.
* Added an optional nullability report to smithy-build that shows the
  computed nullability semantics of each member in a model. This can be
  used to better understand nullability semantics.
* Updated smithy-diff to not emit events when diffing a 1.0 model against
  the 2.0 serialized version of the model. This means that changes to the
  box trait are ignored unless the change impacts the nullability of the
  shape. Special handling was added to detect breaking changes with the
  default trait too (you can't change a default value of a root-level shape
  for example, you can't change a default value of a shape to or from the
  zero value of a type as this might break code generators, etc).
* Add new NullableIndex modes for testing if a member is nullable based on
  the supported features of the generator. For example, some generators only
  make members non-optional when the member is set to the zero value of a
  type, so there is a NullableIndex check mode for that and other use cases.
* Added the @addedDefault trait which is used to indicate that a `@default`
  trait was added to a member after it was initially released. This can be
  used by tooling to make an appropriate determination if generating a
  non-nullable type for the member is a backward compatible change. For
  example, if a generator only uses default zero values to generate
  non-nullable types, then the removal of the required trait and addition
  of a default trait would be a breaking change for them, so they can use
  addedDefault to ignore the default trait.
* When loading 1.0 models, rather than dropping the default trait from a
  member when the range trait of a shape is invalid for its zero value, we
  now instead emit only a warning for this specific case. This prevents
  changing the type and also doesn't lose the range constraint in case
  implementations are relying on this validation.
* Un-deprecated Primitive* shapes in the prelude and add the `@default`
  trait to them. Any member that targets one of these primitive prelude
  shapes must now also repeat the zero value of the target shape.
* Fixed an issue where synthetic box traits were being dropped when the
  same model is loaded multiple times.

### CODE REVIEW

File: smithy-model/src/main/java/software/amazon/smithy/model/traits/TagsTrait.java [hunk 1]
```diff
@@ -37,6 +37,11 @@ public Provider() {
         }
     }
 
+    @Override
+    public List<String> getTags() {
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/traits/TagsTrait.java:41]: This was broken before and needed a fix to make this PR work

**Review by gosar — APPROVED** [review]
All my comments are non-blocking, so marking Approved.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelUpgrader.java [hunk 2]
```diff
@@ -88,94 +114,154 @@ ValidatedResult<Model> transform() {
...
-        // Min is greater than 0.
-        if (rangeTrait.getMin().isPresent() && rangeTrait.getMin().get().compareTo(BigDecimal.ZERO) > 0) {
-            events.add(ValidationEvent.builder()
-                    .id(Validator.MODEL_DEPRECATION)
-                    .severity(Severity.WARNING)
-                    .shape(member)
-                    .message("Cannot add the @default trait to this member due to a minimum range constraint.")
-                    .build());
+    private boolean isDefaultPayload(MemberShape member, Shape target) {
+        // httpPayload requires that the member is required or default.
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelUpgrader.java:143]: yeah looks like the code comment was an old carry-over

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/node/NodeValidatorPlugin.java [hunk 3]
```diff
@@ -108,5 +116,15 @@ public Model model() {
...
+
+        public boolean hasFeature(NodeValidationVisitor.Feature feature) {
+            return features.contains(feature);
+        }
+    }
+
+    @SmithyInternalApi
+    @FunctionalInterface
+    interface Emitter {
+        void accept(FromSourceLocation sourceLocation, Severity severity, String message);
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/validation/node/NodeValidatorPlugin.java:128]: Yeah that can remove a lot of it. done.

File: smithy-cli/src/test/resources/software/amazon/smithy/cli/commands/upgrade/cases/box.v1.smithy [hunk 4]
```diff
@@ -1,39 +0,0 @@
-$version: "1.0"
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-cli/src/test/resources/software/amazon/smithy/cli/commands/upgrade/cases/box.v1.smithy]: Ah yeah also needed to make changes to support default on root shapes. Done,

File: smithy-build/src/test/resources/software/amazon/smithy/build/plugins/nullability/example.smithy [hunk 5]
```diff
@@ -0,0 +1,20 @@
+$version: "1.0"
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-build/src/test/resources/software/amazon/smithy/build/plugins/nullability/example.smithy:1]: Let's circle back to add this

File: smithy-cli/src/test/resources/software/amazon/smithy/cli/commands/upgrade/cases/primitive.v2.smithy [hunk 6]
```diff
@@ -22,7 +22,9 @@ structure PrimitiveBearer {
     handlesComments: PrimitiveShort, // comment
 
     @required
+    @default(0)
     handlesRequired: PrimitiveLong,
 
+    @default(null)
```
Comment (gosar) [inline smithy-cli/src/test/resources/software/amazon/smithy/cli/commands/upgrade/cases/primitive.v2.smithy:28]: hmm, I missed this case earlier. good you caught and fixed it.

File: smithy-diff/src/main/java/software/amazon/smithy/diff/evaluators/ChangedDefault.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            ChangedShape<Shape> change,
+            DefaultTrait oldTrait,
+            DefaultTrait newTrait
+    ) {
+        // Removing the default trait is always a breaking change.
+        if (newTrait == null) {
+            events.add(error(change.getNewShape(),
+                             "@default trait was removed. This will break previously generated code."));
+        } else if (oldTrait == null) {
+            if (change.getNewShape().getType() != ShapeType.MEMBER) {
```
Comment (gosar) [inline smithy-diff/src/main/java/software/amazon/smithy/diff/evaluators/ChangedDefault.java:67]: nit: can also be changed to `!change.getNewShape().isMemberShape()`

**Pull request merged**